python/ray/data/iterator.py

```diff
@@ -158,54 +158,105 @@ def iter_batches(
+                        batch_block_refs(
+                            drop_metadata(block_iterator),
+                            stats=self.stats,
+                            prefetch_blocks=prefetch_blocks,
+                            clear_block_after_read=blocks_owned_by_consumer,
+                            batch_size=batch_size,
+                            batch_format=batch_format,
+                            drop_last=drop_last,
+                            collate_fn=collate_and_finalize,
+                            shuffle_buffer_min_size=local_shuffle_buffer_size,
+                            shuffle_seed=local_shuffle_seed,
+                        )
+                    )
+                else:
+                    return iter(
+                        iter_batches(
+                            block_iterator,
+                            stats=self.stats,
+                            clear_block_after_read=blocks_owned_by_consumer,
+                            batch_size=batch_size,
+                            batch_format=batch_format,
+                            drop_last=drop_last,
+                            collate_fn=_collate_fn,
+                            finalize_fn=_finalize_fn,
+                            shuffle_buffer_min_size=local_shuffle_buffer_size,
+                            shuffle_seed=local_shuffle_seed,
+                            prefetch_batches=prefetch_batches,
+                        )
+                    )
+
+            def __iter__(self):
+                # Iterate through the dataset from the start each time iter is
+                # called.
+                # This allows multiple iterations of the dataset without needing to
+                # explicitly call `iter_batches()` multiple times.
+                self.time_start = time.perf_counter()
+                self.iterator = self._create_iterator()
+                return self
+
+            def set_output_fn(self, fn: Callable[[DataBatch], Any], **kwargs):
+                """Sets a fn to apply to each DataBatch before returning."""
+                self.output_fn = fn
+                self.kwargs = kwargs
+
+            def __next__(self):
+                try:
+                    next_batch = next(self.iterator)
+                    if self.output_fn is not None:
+                        next_batch = self.output_fn(next_batch, **self.kwargs)
+                    return next_batch
+                except StopIteration as e:
+                    if self.stats:
+                        self.stats.iter_total_s.add(
+                            time.perf_counter() - self.time_start
+                        )
+                    raise e
 
-        if stats:
-            stats.iter_total_s.add(time.perf_counter() - time_start)
+        block_iterator_fn = self._to_block_iterator
+        return _BatchIterator(block_iterator_fn)
 
     def iter_rows(self, *, prefetch_blocks: int = 0) -> Iterator[Dict[str, Any]]:
```

Comment:
Update type hints?
```suggestion
    def iter_rows(self, *, prefetch_blocks: int = 0) -> Iterable[Dict[str, Any]]:
```

Reply:
this is still an Iterator

Reply:
well I guess it's both

Reply:
By formal definition, all iterators are also iterables.

Reply:
updated